## Pandas III Metodos

In [1]:
import pandas as pd

1) Cargar el dataset (Pista Si hacemos la carga del dataset sin el parámetro encoding nos puede dar error (puede que no os de error, si es así menos problemas!). Aquí tenéis una lista de los encodings disponibles en pandas. Buscad por los encoding de tipo ISO.)

In [2]:
df = pd.read_csv('attacks.csv')

2) Haced una exploración básica del DataFrame:

- ¿Cuántas filas y columnas tiene el DataFrame?

In [3]:
df.shape

(6114, 22)

Nuestro DataFrame tiene 6114 filas y 22 columnas.

- Mostrad las primeras cinco filas y las últimas cinco filas del DataFrame.

In [4]:
df.head()

Unnamed: 0 Case Number    Year        Type    Country             Area  \
0           0  2018.06.25  2018.0     Boating        USA       California   
1           1  2018.06.18  2018.0  Unprovoked        USA          Georgia   
2           2  2018.06.09  2018.0     Invalid        USA           Hawaii   
3           3  2018.06.08  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4           4  2018.06.04  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   ...  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F  ...   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F  ...   
2                    Habush, Oahu      Surfing      John Denges    M  ...   
3              Arrawarra Headland      Surfing             male    M  ...   
4                        La Ticla  Free diving   Gustavo Ramos     M  ...   

          Species           Investigator or Source                       pdf  \
0      White shark                R. Collier, GSAF      2018.06.25-Wolfe.pdf   
1              NaN  K.McMurray, TrackingSharks.com    2018.06.18-McNeely.pdf   
2              NaN  K.McMurray, TrackingSharks.com     2018.06.09-Denges.pdf   
3        2 m shark                  B. Myatt, GSAF  2018.06.08-Arrawarra.pdf   
4  Tiger shark, 3m                       A .Kipper      2018.06.04-Ramos.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.04   

  Case Number.2 original order Unnamed: 22  Unnamed: 23  
0    2018.06.25         6303.0         NaN          NaN  
1    2018.06.18         6302.0         NaN          NaN  
2    2018.06.09         6301.0         NaN          NaN  
3    2018.06.08         6300.0         NaN          NaN  
4    2018.06.04         6299.0         NaN          NaN  

[5 rows x 22 columns]

In [5]:
df.tail()

Unnamed: 0   Case Number    Year          Type    Country  \
6109        6127  1804.02.26.R  1804.0          Boat  AUSTRALIA   
6110        6128    1803.05.17  1803.0  Sea Disaster        USA   
6111        6129    1803.03.00  1803.0    Unprovoked  AUSTRALIA   
6112        6130  1802.04.13.R  1802.0    Unprovoked      INDIA   
6113        6131  1801.12.18.R  1801.0      Provoked        NaN   

                   Area                          Location  \
6109    New South Wales    Georges Head, off Port Jackson   
6110     South Carolina                    Off Charleston   
6111  Western Australia  Hamelin Harbour, at Faure Island   
6112                NaN                               NaN   
6113                NaN                               NaN   

                             Activity                             Name Sex   \
6109                              NaN                             boat  NaN   
6110                              NaN                    Captain Jones    M   
6111                              NaN  M. Lefevre & a sailor (rescuer)    M   
6112                              NaN                              NaN  NaN   
6113  Standing on landed shark's tail                 Stephen Pettigew    N   

      ...   Species                              Investigator or Source  \
6109  ...        NaN                         Sydney Gazette, 2/26/1804    
6110  ...        NaN                            Evening Post, 6/13/1803   
6111  ...        NaN  F.  Peron ref in G.P. Whitley (Fishes of Austr...   
6112  ...        NaN                  Evening Post (New York) 4/13/1802   
6113  ...  12' shark                       The Evening Post, 12/18/1801   

                                  pdf  \
6109  1804.02.26-boat-GeorgesHead.pdf   
6110             1803.05.17-Jones.pdf   
6111           1803.03.00-Lefevre.pdf   
6112           1802.04.13.R-India.pdf   
6113       1801.12.18.R-Pettigrew.pdf   

                                           href formula  \
6109  http://sharkattackfile.net/spreadsheets/pdf_di...   
6110  http://sharkattackfile.net/spreadsheets/pdf_di...   
6111  http://sharkattackfile.net/spreadsheets/pdf_di...   
6112  http://sharkattackfile.net/spreadsheets/pdf_di...   
6113  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href Case Number.1  \
6109  http://sharkattackfile.net/spreadsheets/pdf_di...  1804.02.26.R   
6110  http://sharkattackfile.net/spreadsheets/pdf_di...    1803.05.17   
6111  http://sharkattackfile.net/spreadsheets/pdf_di...    1803.03.00   
6112  http://sharkattackfile.net/spreadsheets/pdf_di...  1802.04.13.R   
6113  http://sharkattackfile.net/spreadsheets/pdf_di...  1801.12.18.R   

     Case Number.2 original order Unnamed: 22  Unnamed: 23  
6109  1804.02.26.R          176.0         NaN          NaN  
6110    1803.05.17          175.0         NaN          NaN  
6111    1803.03.00          174.0         NaN          NaN  
6112  1802.04.13.R          173.0         NaN          NaN  
6113  1801.12.18.R          172.0         NaN          NaN  

[5 rows x 22 columns]

- Mostrad 2 filas al azar del DataFrame.

In [6]:
df.sample(2)

Unnamed: 0     Case Number    Year        Type Country             Area  \
5642        5659  1894.06.15.a.R  1894.0  Unprovoked  MEXICO  Bay of Campeche   
362          363      2015.09.17  2015.0  Unprovoked     USA          Florida   

                              Location  \
5642                               NaN   
362   Jacksonville Beach, Duval County   

                                     Activity             Name Sex   ...  \
5642  Fell from yardarm of British ship Rover           sailor    M  ...   
362                                   Surfing  Bryan Liebetrau    M  ...   

       Species     Investigator or Source                       pdf  \
5642  20' shark  Warren Ledger, 6/15/1894  1894.06.15.a-R-Rover.pdf   
362         NaN       News4JAX, 9/15/2015  2015.09.17-Liebetrau.pdf   

                                           href formula  \
5642  http://sharkattackfile.net/spreadsheets/pdf_di...   
362   http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href   Case Number.1  \
5642  http://sharkattackfile.net/spreadsheets/pdf_di...  1894.06.15.a.R   
362   http://sharkattackfile.net/spreadsheets/pdf_di...      2015.09.17   

       Case Number.2 original order Unnamed: 22  Unnamed: 23  
5642  1894.06.15.a.R          644.0         NaN          NaN  
362       2015.09.17         5940.0         NaN          NaN  

[2 rows x 22 columns]

- ¿Cuál es el nombre de las columnas del DataFrame?

In [7]:
columnas= df.columns
columnas

Index(['Unnamed: 0', 'Case Number', 'Year', 'Type', 'Country', 'Area',
       'Location', 'Activity', 'Name', 'Sex ', 'Age', 'Time', 'Species ',
       'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23'],
      dtype='object')

- ¿Cuáles son los principales estádisticos del DataFrame? Mostrad por separado las variables categóricas y las variables numéricas.

In [8]:
df.describe() # Variables numericas

Unnamed: 0         Year  original order
count  6114.000000  6114.000000     6114.000000
mean   3065.554629  1968.831207     3237.445044
std    1769.652139    43.745920     1769.652633
min       0.000000  1801.000000      172.000000
25%    1534.250000  1946.000000     1706.250000
50%    3066.500000  1980.000000     3236.500000
75%    4596.750000  2005.000000     4768.750000
max    6131.000000  2018.000000     6303.000000

In [9]:
df.describe(include= 'object') # Variables categoricas

Case Number        Type Country     Area  \
count         6113        6110    6074     5706   
unique        6113           8     201      791   
top     2018.06.25  Unprovoked     USA  Florida   
freq             1        4438    2200     1025   

                                Location Activity  Name  Sex    Age  \
count                               5626     5603  5914  5564  3448   
unique                              4020     1484  5118     6   156   
top     New Smyrna Beach, Volusia County  Surfing  male     M    17   
freq                                 162      966   511  4939   154   

             Time     Species  Investigator or Source                     pdf  \
count        2934         3409                   6098                    6114   
unique        366         1529                   4837                    6105   
top     Afternoon  White shark         C. Moore, GSAF  1898.00.00.R-Syria.pdf   
freq          183          161                     95                       2   

                                             href formula  \
count                                                6113   
unique                                               6104   
top     http://sharkattackfile.net/spreadsheets/pdf_di...   
freq                                                    2   

                                                     href Case Number.1  \
count                                                6114          6114   
unique                                               6099          6113   
top     http://sharkattackfile.net/spreadsheets/pdf_di...    1952.08.04   
freq                                                    4             2   

       Case Number.2   Unnamed: 22 Unnamed: 23  
count           6114             1           2  
unique          6114             1           2  
top       2018.06.25  stopped here      Teramo  
freq               1             1           1

- ¿Cuál es el porcentaje de valores nulos en cada columna?

In [21]:
for i in columnas:
    resultado = (df[i].isnull().sum()/len(df[i]))*100
    print(f"{i} {round(resultado,2)} %")

Unnamed: 0 0.0 %
Case Number 0.02 %
Year 0.0 %
Type 0.07 %
Country 0.65 %
Area 6.67 %
Location 7.98 %
Activity 8.36 %
Name 3.27 %
Sex  9.0 %
Age 43.6 %
Time 52.01 %
Species  44.24 %
Investigator or Source 0.26 %
pdf 0.0 %
href formula 0.02 %
href 0.0 %
Case Number.1 0.0 %
Case Number.2 0.0 %
original order 0.0 %
Unnamed: 22 99.98 %
Unnamed: 23 99.97 %


- Sacad los valores únicos y su frecuencia de las variables numéricas. Almacenad los resultados en un diccionario. Uno para cada variable.

In [22]:
dic_unicos ={} #Creamos un diccionario con variables numericas
for i in df:
    
    if df.dtypes[i] == 'object':
        pass
    else:
         dic_unicos.update({i:(df[i].value_counts())})

In [37]:
dic_unicos.keys() #Verificamos que columnas trajo

dict_keys(['Unnamed: 0', 'Year', 'original order'])

In [42]:
# convertimos el diccionario en tres uno por cada variable numerica como pedia la consigna.

dic_unnamed = dict(dic_unicos['Unnamed: 0']) 
dic_year = dict(dic_unicos['Year'])
dic_original = dict(dic_unicos['original order'])


- Cambiad el nombre de las columnas para que todas estén en minúsculas y reemplazad los espacios por _.

In [45]:
cambio_columnas = {col :  col.lower().strip().replace(' ','_') for col in df.columns}

In [46]:
df1 = df.rename(columns = cambio_columnas) #inplace = True
df1

unnamed:_0   case_number    year          type    country  \
0              0    2018.06.25  2018.0       Boating        USA   
1              1    2018.06.18  2018.0    Unprovoked        USA   
2              2    2018.06.09  2018.0       Invalid        USA   
3              3    2018.06.08  2018.0    Unprovoked  AUSTRALIA   
4              4    2018.06.04  2018.0      Provoked     MEXICO   
...          ...           ...     ...           ...        ...   
6109        6127  1804.02.26.R  1804.0          Boat  AUSTRALIA   
6110        6128    1803.05.17  1803.0  Sea Disaster        USA   
6111        6129    1803.03.00  1803.0    Unprovoked  AUSTRALIA   
6112        6130  1802.04.13.R  1802.0    Unprovoked      INDIA   
6113        6131  1801.12.18.R  1801.0      Provoked        NaN   

                   area                          location  \
0            California       Oceanside, San Diego County   
1               Georgia    St. Simon Island, Glynn County   
2                Hawaii                      Habush, Oahu   
3       New South Wales                Arrawarra Headland   
4                Colima                          La Ticla   
...                 ...                               ...   
6109    New South Wales    Georges Head, off Port Jackson   
6110     South Carolina                    Off Charleston   
6111  Western Australia  Hamelin Harbour, at Faure Island   
6112                NaN                               NaN   
6113                NaN                               NaN   

                             activity                             name  sex  \
0                            Paddling                      Julie Wolfe    F   
1                            Standing                  Adyson McNeely     F   
2                             Surfing                      John Denges    M   
3                             Surfing                             male    M   
4                         Free diving                   Gustavo Ramos     M   
...                               ...                              ...  ...   
6109                              NaN                             boat  NaN   
6110                              NaN                    Captain Jones    M   
6111                              NaN  M. Lefevre & a sailor (rescuer)    M   
6112                              NaN                              NaN  NaN   
6113  Standing on landed shark's tail                 Stephen Pettigew    N   

      ...          species                             investigator_or_source  \
0     ...      White shark                                   R. Collier, GSAF   
1     ...              NaN                     K.McMurray, TrackingSharks.com   
2     ...              NaN                     K.McMurray, TrackingSharks.com   
3     ...        2 m shark                                     B. Myatt, GSAF   
4     ...  Tiger shark, 3m                                          A .Kipper   
...   ...              ...                                                ...   
6109  ...              NaN                         Sydney Gazette, 2/26/1804    
6110  ...              NaN                            Evening Post, 6/13/1803   
6111  ...              NaN  F.  Peron ref in G.P. Whitley (Fishes of Austr...   
6112  ...              NaN                  Evening Post (New York) 4/13/1802   
6113  ...        12' shark                       The Evening Post, 12/18/1801   

                                  pdf  \
0                2018.06.25-Wolfe.pdf   
1              2018.06.18-McNeely.pdf   
2               2018.06.09-Denges.pdf   
3            2018.06.08-Arrawarra.pdf   
4                2018.06.04-Ramos.pdf   
...                               ...   
6109  1804.02.26-boat-GeorgesHead.pdf   
6110             1803.05.17-Jones.pdf   
6111           1803.03.00-Lefevre.pdf   
6112           1802.04.13.R-India.pdf   
6113       1801.12.18.R-Pettigrew.pdf   

                                           href_formu

In [47]:
df1.columns  # verificamos los cambios

Index(['unnamed:_0', 'case_number', 'year', 'type', 'country', 'area',
       'location', 'activity', 'name', 'sex', 'age', 'time', 'species',
       'investigator_or_source', 'pdf', 'href_formula', 'href',
       'case_number.1', 'case_number.2', 'original_order', 'unnamed:_22',
       'unnamed:_23'],
      dtype='object')

- Eliminad las columnas de Unnamed: 22 y Unnamed: 23, Case Number.1, Case Number.2, original order, Investigator or Source, pdf, href formula.

In [48]:
df1.drop(['investigator_or_source', 'pdf', 'href_formula','case_number.1', 'case_number.2', 'original_order', 'unnamed:_22',
       'unnamed:_23'], axis=1, inplace=True)

In [49]:
df1.columns # Verificamos los cambios

Index(['unnamed:_0', 'case_number', 'year', 'type', 'country', 'area',
       'location', 'activity', 'name', 'sex', 'age', 'time', 'species',
       'href'],
      dtype='object')

- Establece la columna Case Number como índice del DataFrame.

In [52]:
df1.set_index(['case_number'], inplace= True)
df1.head(1)

unnamed:_0    year     type country        area  \
case_number                                                    
2018.06.25            0  2018.0  Boating     USA  California   

                                location  activity         name sex age  \
case_number                                                               
2018.06.25   Oceanside, San Diego County  Paddling  Julie Wolfe   F  57   

              time      species  \
case_number                       
2018.06.25   18h00  White shark   

                                                          href  
case_number                                                     
2018.06.25   http://sharkattackfile.net/spreadsheets/pdf_di...

- Guarda el csv que no tiene esas columnas para seguir usándolo mañana.

In [53]:
df1.to_csv('attacks1.csv')